lib/lightstep/tracer/tracer.rb

```diff
@@ -0,0 +1,374 @@
+        tags: tags
+      )
+    end
+
+    # Inject a span into the given carrier
+    # @param span [Span]
+    # @param format [LightStep::Tracer::FORMAT_TEXT_MAP, LightStep::Tracer::FORMAT_BINARY]
+    # @param carrier [Hash-like]
+    def inject(span, format, carrier)
+      case format
+      when LightStep::Tracer::FORMAT_TEXT_MAP
+        inject_to_text_map(span, carrier)
+      when LightStep::Tracer::FORMAT_BINARY
+        warn 'Binary inject format not yet implemented'
+      else
+        warn 'Unknown inject format'
+      end
+    end
+
+    # Extract a span from a carrier
+    # @param operation_name [String]
+    # @param format [LightStep::Tracer::FORMAT_TEXT_MAP, LightStep::Tracer::FORMAT_BINARY]
+    # @param carrier [Hash-like]
+    # @return [Span]
+    def extract(operation_name, format, carrier)
+      case format
+      when LightStep::Tracer::FORMAT_TEXT_MAP
+        extract_from_text_map(operation_name, carrier)
+      when LightStep::Tracer::FORMAT_BINARY
+        warn 'Binary join format not yet implemented'
+        nil
+      else
+        warn 'Unknown join format'
+        nil
+      end
+    end
+
+    # @return true if the tracer is enabled
+    def enabled?
+      @enabled ||= true
+    end
+
+    # Enables the tracer
+    def enable
+      @enabled = true
+    end
+
+    # Disables the tracer
+    # @param discard [Boolean] whether to discard queued data
+    def disable(discard: true)
+      @enabled = false
+      @tracer_transport.clear if discard
+      @tracer_transport.flush
+    end
+
+    # Flush to the Transport
+    def flush
+      _flush_worker
+    end
+
+    # Internal use only.
+    # @private
+    def _finish_span(span, end_time: Time.now)
+      return unless enabled?
+
+      span.end_micros ||= micros(end_time)
+      full = push_with_max(@tracer_span_records, span.to_h, max_span_records)
+      @dropped_spans.increment if full
+      flush_if_needed
+    end
+
+    # Internal use only
+    # @private
+    def raw_log_record(fields, payload)
+      return unless enabled?
+
+      fields['runtime_guid'] = guid
+
+      if fields['timestamp_micros'].nil?
+        fields['timestamp_micros'] = now_micros
+      end
+
+      # TODO: data scrubbing and size limiting
+      json = nil
+      case payload
+      when Array, Hash
+        begin
+          fields['payload_json'] = JSON.generate(payload, max_nesting: 8)
+        rescue
+          # TODO(ngauthier@gmail.com) naked rescue
+          # TODO: failure to encode a payload as JSON should be recorded in the
+          # internal library logs, with catioun not flooding the internal logs.
+        end
+      when nil
+        # noop
+      else
+        # TODO: Remove the outer 'payload' key wrapper. Just transport the JSON
+        # Value (Value in the sense of the JSON spec).
+        fields['payload_json'] = JSON.generate(payload: payload)
+      end
+
+      full = push_with_max(@tracer_log_records, fields, max_log_records)
```

Comment:
one conceptual change is that logs should *always* be associated with a span... in other language libs they are literally stored within the `Span` instance somehow, and we ought to do the same thing here.

it means that `max_log_records` is more like `max_log_records_per_span` or whatever.